src/split.ts

```diff
@@ -0,0 +1,320 @@
+  }
+
+  hasUnEmittedParts(): boolean {
+    // First chunk.
+    if (this.lastEmittedPart === null) return this.parts.length > 0
+
+    // Subsequent chunks.
+    if (this.parts.length > 0) {
+      // Check if have un-emitted parts past the end of last emitted part.
+      const lastPart = this.parts[this.parts.length - 1]
+      return lastPart.end > this.lastEmittedPart.end
+    }
+
+    // Otherwise, we have no un-emitted parts.
+    return false
+  }
+
+  emit(): Chunk {
+    // Sanity check.
+    if (this.parts.length === 0) throw new Error('Chunk parts is empty')
+
+    // Prepare chunk.
+    const start: number = this.parts[0].start
+    this.lastEmittedPart = this.parts[this.parts.length - 1]
+    const end: number = this.lastEmittedPart.end
+    const chunk: Chunk = {
+      text: getChunk(this.input, start, end),
+      start,
+      end
+    }
+
+    // Reset state.
+    // At this point, we seed the new parts array with the overlap, if any found.
+    if (this.chunkOverlap > 0) {
+      this.parts = this.parts.slice(-this.chunkOverlap)
+    } else {
+      this.parts = []
+    }
+
+    // Clear out last boundary. We consider a new chunk to have "no" previous boundary.
+    this.lastBoundaryPart = null
+
+    return chunk
+  }
+}
+
+/**
+ * Split text into chunks.
+ *
+ * ## Chunk Structure
+ * Note that when splitting into tokens if an array is passed to input, the array item boundary is
+ * *always* a token boundary.
+ *
+ * In the returned structure, `start` is the start of the first token in the chunk and `end` is
+ * the end of the last token. In between there may be unmatched / discarded parts between tokens
+ * (e.g. if you split on whitespace, there may be spaces between tokens). The `text` field of
+ * the returned chunk will include all the text or array of texts from the start to the end,
+ * inclusive of the unmatched parts.
+ *
+ * ## Chunk Strategy
+ * The `chunkStrategy` option allows you to specify how the chunks are grouped.
+ * - `character`: There is no grouping preference here. Fit as many whole tokens as possible into a chunk.
+ * - `paragraph`: Group tokens by paragraphs. If a paragraph exceeds the chunk size, it will be split across multiple chunks.
+ *
+ * @param {string|string[]} input - The input (string or array of strings) to split.
+ * @param {Object} options
+ * @param {number} options.chunkSize - The max number of tokens (from splitter) of each chunk.
+ * @param {number} options.chunkOverlap - The overlapping number of tokens (from splitter) to include from previous chunk.
+ * @param {Function} options.splitter - The function to split the text.
+ * @param {string} options.chunkStrategy - The strategy used to group tokens into chunks.
+ * @returns {Array<{text: string | null, start: number, end: number}>}
+ */
+export function split(
+  input: string | string[],
+  {
+    chunkSize = 512,
+    chunkOverlap = 0,
+    splitter = text => text.split(''),
```

Comment:
OPTIONAL -- I was thinking of maybe switching globally to `Array.from(text)` as it handles multibyte characters better. Not sure though...

Reply:
Or even `[...text]`

Reply:
OK! -- Your choice on this one. And `[...text]` is fine (equivalent perf and less characters to type).

If you do change, just make sure to change everywhere (tests, code, docs).

Thanks!